ethcore/src/engines/mod.rs

```diff
@@ -338,6 +340,27 @@ pub trait Engine<M: Machine>: Sync + Send {
 	fn is_timestamp_valid(&self, header_timestamp: u64, parent_timestamp: u64) -> bool {
 		header_timestamp > parent_timestamp
 	}
+
+	/// Generate metadata for a new block.
+	fn generate_metadata(&self, bytes: &[u8], provider: &M::BlockProvider) -> M::BlockMetadata;
+
+	/// Check whether a given block is the best block.
+	fn is_new_best(&self, bytes: &[u8], block_metadata: &M::BlockMetadata, best_block_metadata: &M::BlockMetadata, provider: &M::BlockProvider) -> bool;
+}
+
+/// Generate metadata for a new block based on the default total difficulty rule.
+pub fn total_difficulty_generate_metadata(bytes: &[u8], provider: &BlockProvider) -> U256 {
+	let block = BlockView::new(bytes);
+	let header = block.header_view();
+	let parent_hash = header.parent_hash();
+	let parent_details = provider.block_details(&parent_hash).unwrap_or_else(|| panic!("Invalid parent hash: {:?}", parent_hash));
```

Comment:
it would simplify things a lot just to provide the parent metadata

Reply:
> we now have logic in engine that needs to check a block's timestamp

it probably should be abstracted at some point. we have `verify_block`/`verify_header` functions that can handle it AFAIK.

> InternalSealer, BlockVerifier, TimestampVerifier, ForkChooser

The sub-traits approach would work if we weren't using Engine as a trait object everywhere. That's why there's the `EthEngine` specialization which is implementation for all `Engine<EthereumMachine>`.

I don't see anything in this PR that requires introduction of a dependency of Ethereum types.
In the parity-machine crate we could add an associated type `Metadata` and a 

```rust
trait TotalDifficulty {
    type Value: Ord;

    fn total_difficulty(&self) -> &Value;
}
```

and then implement this trait for `U256` or some wrapper in the `EthereumMachine` implementation. Our engines then simply need the bound that `M::Metadata: TotalDifficulty`.